src/java/org/apache/cassandra/db/memtable/TrieMemtable.java

```diff
@@ -369,16 +370,36 @@ public FlushablePartitionSet<MemtablePartition> getFlushSet(PartitionPosition fr
         long partitionKeySize;
         long partitionCount;
 
-        if (allPositionsToFlush)
+        int fromShard = Integer.MIN_VALUE;
+        int toShard = Integer.MIN_VALUE;
+        if (from == null || from.isMinimum())
+            fromShard = 0;
+
+        if  (to == null || to.isMaxKeyBound())
         {
-            partitionKeySize = this.partitionKeysTotalSize();
-            partitionCount = partitionCount();
+            toShard = boundaries.shardCount() - 1;
         }
-        else
+
+        for (int i = 0; i < boundaries.shardCount() - 1; i++)
         {
-            long keySize = 0;
-            int keyCount = 0;
+            Token shardRightBoundary = boundaries.getShardRightBoundary(i);
+            if (from != null && shardRightBoundary.equals(from.getToken()))
```

Comment:
Because `from` and `to` are `PartitionPosition`, they could be different from this boundary even if this test returns true (e.g. they could be a specific `DecoratedKey` instead of a token, or they could be a token min bound). Because the boundaries are end-inclusive, we should be converting the right boundary via `Token.maxKeyBound` and comparing that against `from`/`to` rather than its token.